### PR TITLE
Drag and Drop for Admin only, cleanup AgendaView.

### PR DESCRIPTION
**Pull request by ppelayo1**
### :scroll: Description

- Describe the general shape of this PR (new feature? refactor? bug fix? one-line change?)
This pull request will update the front-end so that only admin's can sort agenda items, and that such an interface is only visible to admins.
Also now renders the agenda items by their order number. By default this is set to 1, but can be changed thru an option variable in AgendaView.jsx . 
Also removes the ability to sort the agenda containers themselves as that was an unwanted feature.
Lastly the code for AgendaView.jsx was cleaned up and simplified. 

- Describe what changes are being made

**Code Cleanup**
Took many of the functions in AgendaView.jsx and put into agendaViewFunctions.
The other folder is DragOverlayHandlers
This is a component to display the drag and drop overlay. That is when the user begins dragging a item it displays a identical item that is dragged around by the user.

createRenderedGroups.jsx extracted from AgendaView.jsx into its own file

dndKitFunctions.jsx extracted from AgendaView.jsx into its own file. This holds the event handlers when a user is dragging and dropping. It included changes to remove the ability to sort agendaGroups and line 40 resorts the order number of all agenda items when performing a drag and drop.

groupMeetingItems.jsx extracted from AgendaView.jsx into its own file. 

**changes and new things**

saveReOrder.jsx this function will perform the graphql mutations to save the meeting items

MeetingView.jsx changed so that it can pass a flag and its setter between Header.jsx and AgendaView.jsx. This tells AgendaView.jsx to perform a save of the meeting items. Then it resets the flag.

MeetingView.jsx is also updated to perform a fresh query on the meeting items when a user clicks participate. When this is done a timer is set to perform that query, and this time can be configured in an OPTIONS variable. This is necessary as if the ADMIN saves the data and then clicks participate too quickly the query to fetch the meeting items would execute before the mutator in AgendaView.jsx would have finished executing and cause stale data when the ADMIN switches back to agenda
AgendaView.jsx  calls the saveReOrder function when update status: is clicked in the header. 

AgendaView.jsx updated also to rebuild the AgendaGroups state whenever meetings propr passed by MeetingView.jsx changes. Necessary to account for switching between particpate and agenda buttons.

AgendaGroups.jsx removed the ability to sort Agenda group containers, and changed the component so that it will not permit sorting when not an ADMIN. This is done at line 107.

UserRoles.js a constant Object that contains the roles ADMIN and USER.

graphql.js added UpdateMeetingItem mutation to the front-end.

getAdminStatus.js utility function that can be used to get a true or false on if the current user is a ADMIN.

- Describe why these changes are being made
#219
#209
#199
#198
#197

- List the use cases and edge cases relevant to this PR
Regular use case would just be dragging and sorting agenda items and then hitting save.
Edge case would be the timer in MeetingView.jsx executing before the meeting items were updated in the AgendaView.jsx mutator. This would cause stale data.

- Describe how errors will be handled. How will we know if this code breaks in production

- Describe any external libraries/dependencies added or removed in this PR
none

- Describe any security risks are there regarding this change

- Describe how you tested these changes
I used `<p style={{ color: 'black', fontWeight: 900 }}>
        Order Number is
        {' '}
        {item.order_number}
      </p>`
on line 127 in AgendaItem.jsx to view the order number. I ensured it was updating correctly. I then tested drag and dropping between agenda groups. Saving, then logging in and out of different accounts, one with ADMIN status and one with only USER. Tested drag and dropping, and then saving the meeting items. Then refreshing page.
- Link to relevant external documentation

--------------------------
### :clipboard: Mandatory Checklist

- [x] Example of a checked item (please remove when creating your Pull Request) 

- [x] Linked to the Github Issues being addressed using the right sidebar :arrow_right:
- [ ] Have you discussed these changes with the project leader(s)?
- [x] Do all variable and function names communicate what they do?
- [x] Were all the changes commented and / or documented?
- [ ] Is the PR the right size? (If the PR is too large to review, it might be good to break it up into multiple PRs.)
- [ ] Does all work in progress, temporary, or debugger code have a TODO comment with links to Github issues?
- [ ] *If you changed the user interface, did you add before and after screenshots to below?*

--------------------------
#### After

https://user-images.githubusercontent.com/32974267/122097880-04efae00-cdc5-11eb-9506-0d63d6c8d5fb.mp4

--------------------------
### :blue_book: Glossary
- PR = Pull Request

### CODE REVIEW

**Review by JMStudiosJoe — DISMISSED** [review]
fantastic refactor and code cleanup thank you!

**Comment by JMStudiosJoe** [comment]
@anna-kartynnik could you please take a look as well, all looks good to me but its a big pr and feel free to merge if you are also good with pr

**Comment by ppelayo1** [comment]
Thank you, I am trying to review my work more for my pull requests so that you don't have to point out errors I should have spot.

File: frontend/src/components/MeetingView/AgendaView/agendaViewFunctions/saveReOrder.jsx [hunk 1]
```diff
@@ -0,0 +1,23 @@
+// saves the new rendering
+const saveReOrder = (agendaGroups, updateMeetingItem) => {
+  for (let parent = 0; parent < agendaGroups.length; parent += 1) {
+    for (let child = 0; child < agendaGroups[parent].items.length; child += 1) {
+      updateMeetingItem({
```
Comment (anna-kartynnik) [inline frontend/src/components/MeetingView/AgendaView/agendaViewFunctions/saveReOrder.jsx:5]: Here you update all the meeting items making one request to the server per item. This is very inefficient. First of all, not all the items need to be updated. It is possible only two of them changed the order, then only those two need to be saved. Secondly, I would suggest asking a backend team to create a new endpoint for saving items order in a batch. This way only one request to the server will do the work.

File: frontend/src/components/MeetingView/MeetingView.jsx [hunk 2]
```diff
@@ -24,35 +24,74 @@ import Spinner from '../Spinner/Spinner';
  *    navToggled
  *      A boolean value indicating if the header nav component is open
  */
+const OPTIONS = {
+  refreshTimer: 1000, // ms to refresh meetingWithItems after clicking participate
+};
+
+const createMeetings = (data, setMeetingWithItems) => {
```
Comment (anna-kartynnik) [inline frontend/src/components/MeetingView/MeetingView.jsx:31]: It seems this function works with one meeting only. Rename to `createMeeting`?

File: frontend/src/components/MeetingView/MeetingView.jsx [hunk 3]
```diff
@@ -24,35 +24,74 @@ import Spinner from '../Spinner/Spinner';
...
+};
 
 function MeetingView() {
   const { t } = useTranslation();
   const { id } = useParams();
 
+  // queries
   const { loading, error, data } = useQuery(GET_MEETING_WITH_ITEMS, {
     variables: { id: parseInt(id, 10) },
+    fetchPolicy: 'network-only',
```
Comment (anna-kartynnik) [inline frontend/src/components/MeetingView/MeetingView.jsx:44]: What do you think about using `cache-and-network` policy here?

**Review by anna-kartynnik — CHANGES_REQUESTED** [review]
Thank you for refactoring! It is a great work. Please, don't take my comments as I point to your errors. There are just my thoughts. They can be wrong. I think, PR is a great place to discuss some decisions. It is always interesting to know how other people would solve the same problem.

There are some minor issues/comments. You can find them below. Some other remarks: you don't need to have `.jsx` extension in the file if it doesn't contain JavaScript XML.

The solution with timer seems odd to me. Why 1 second should be enough? What if I have a very slow network, do I need 5 seconds? I think a better solution would be to disable Participate tab until we get a response from the server. We can even show an additional modal window with the message like `Your changes are being saved` and a spinner. This modal window can block the page, so user can't click anything else. Of course, such a modal window needs to be confirmed by the designers and Annie.

I also think that this is very ineffective to update meeting items one by one. Some of them have nothing changed, so we shouldn't touch them at all. But even if all of them have been changed, ideally, we should do only 1 request to the server with all the items that need update.

File: frontend/src/components/MeetingView/MeetingView.jsx (hunk 3 above)
Comment (ppelayo1) [reply frontend/src/components/MeetingView/MeetingView.jsx:44]: I was having problems with the cache loading despite making updates thru the mutate. I will try cache-and-network again to see if maybe the timer, or response approach solved the issue.

File: frontend/src/components/MeetingView/MeetingView.jsx (hunk 2 above)
Comment (ppelayo1) [reply frontend/src/components/MeetingView/MeetingView.jsx:31]: Good point.

File: frontend/src/components/MeetingView/AgendaView/agendaViewFunctions/saveReOrder.jsx (hunk 1 above)
Comment (ppelayo1) [reply frontend/src/components/MeetingView/AgendaView/agendaViewFunctions/saveReOrder.jsx:5]: I will look into this. You are absolutely right it is inefficient the way I did it.

File: frontend/src/components/Header/Header.scss [hunk 4]
```diff
@@ -63,6 +63,23 @@
...
+        background-color:$cityBlue;
+        color:white;
+        border-style:inset;
+        margin-left:0.7rem;
+        padding-right:1.4rem;
+        text-align: left;
+        cursor:pointer;
+      }
+
+      .saveStatusButton:active{
```
[the remark this answers is not in the document]
Comment (ppelayo1) [reply frontend/src/components/Header/Header.scss:78]: You're right, I screwed up there. Somehow forgot to add the braces.

File: frontend/src/components/Header/Header.jsx [hunk 5]
```diff
@@ -62,6 +67,20 @@ function Header({ loading, meeting }) {
...
               </div>
+
+              {
+              getAdminStatus() && showUpdateStatus && (
+                <>
+                  <div className="saveStatus">
+                    {t('meeting.status.label')}
+                    :
+                    <button className="saveStatusButton" type="button" onClick={() => { setSaveMeetingItems(true); }}>
+                      {t('meeting.status.short.default-option')}
```
[the remark this answers is not in the document]
Comment (ppelayo1) [reply frontend/src/components/Header/Header.jsx:78]: Yeah, I wasn't sure where else to put the save button.

File: frontend/src/utils/getAdminStatus.js [hunk 6]
```diff
@@ -0,0 +1,13 @@
+import parseJwt from './parseJwt';
+import LocalStorageTerms from '../constants/LocalStorageTerms';
+import UserRoles from '../constants/UserRoles';
+
+// returns true if the user is an admin
+const getAdminStatus = () => {
```
[the remark this answers is not in the document]
Comment (ppelayo1) [reply frontend/src/utils/getAdminStatus.js:6]: Thank you for that suggestion. I believe you are right and that is a better practice.

**Comment by ppelayo1** [comment]
I greatly appreciate the feedback. I will probably take a few days before I can get to work on making the changes to improve the code.

I have never worked an engineering job , so I don't want to create extra work for your reviews due to my inexperience creating simple mistakes that I should catch.

**Comment by JMStudiosJoe** [comment]
no worries @ppelayo1 we greatly appreciate your work and no worries contribute when you can. PR reviews are the best way to learn and grow from so no worries on any mistakes please fix them when you can and so far you are doing great engineering work :) We can also connect and go over these changes together. it may seem like a lot but that is how refactors just go sometimes and the more people know about these the better. So please if you need to reach out to other devs for helping with this review changes then let us know. :) thanks!

**Comment by JMStudiosJoe** [comment]
hello @ppelayo1 would you be able to meet sometime later this week to wrap up this pr?

**Comment by ppelayo1** [comment]
I am gonna get started working on it again tomorrow.

File: frontend/src/components/MeetingView/AgendaView/agendaViewFunctions/createRenderedGroups.jsx [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+  for (let i = 0; i < newAgendaGroups.length; i += 1) {
+    if (newAgendaGroups[i].status !== MeetingItemStates.COMPLETED) {
+      const group = newAgendaGroups[i];
+      uncompletedOnly.push(group);
+      const items = group.items.filter((item) => item.status !== MeetingItemStates.COMPLETED);
+      newAgendaGroups[i].items = items;
+    }
+  }
+
+  return (showCompleted ? newAgendaGroups : uncompletedOnly);
```
[the remark this answers is not in the document]
Comment (ppelayo1) [reply frontend/src/components/MeetingView/AgendaView/agendaViewFunctions/createRenderedGroups.jsx]: Turns out I didn't even need line 18 or the showCompleted variable as AgendaView.jsx line 155 already was performing the check before calling the function.

File: frontend/src/components/MeetingView/AgendaView/agendaViewFunctions/saveReOrder.jsx (hunk 1 above)
Comment (ppelayo1) [reply frontend/src/components/MeetingView/AgendaView/agendaViewFunctions/saveReOrder.jsx:5]: I agree a mutator that will allow updating multiple items in one batch would be best. 
I rather not just update only the items changed as then I will have to account for the change in order number of everything where that item used to be and above it, as well as in the new agenda group.

Would be a bunch of unnecessary mutations. Batch mutator is an ideal solution here.

Comment (ppelayo1) [reply frontend/src/components/MeetingView/AgendaView/agendaViewFunctions/saveReOrder.jsx:5]: https://www.apollographql.com/docs/react/api/link/apollo-link-batch-http/ 
Apollo has a way to send in batches without needing to mess with the back end or graphql types.
Did some experimenting.

Looks like it has an option "batchMax" to specify the maximum number to send in a batch, and a second "batchInterval" to specify how long to wait before sending the batch.

This should improve the performance by just sending the first 100 or so mutations as a single batch.

Comment (ppelayo1) [reply frontend/src/components/MeetingView/AgendaView/agendaViewFunctions/saveReOrder.jsx:5]: I am going to look into implementing a way to update thru mutators only the changed items. Would be more efficient even with batch.

Comment (ppelayo1) [reply frontend/src/components/MeetingView/AgendaView/agendaViewFunctions/saveReOrder.jsx]: Only the items changed will now be mutated.

**Comment by ppelayo1** [comment]
This pull request should be more or less ready for merger. Just need to find out if we will have a modal that prevents admin from doing anything during item saves.

File: frontend/src/components/Header/Header.scss (hunk 4 above)
[the remark this answers is not in the document]
Comment (anna-kartynnik) [reply frontend/src/components/Header/Header.scss:77]: You could also add this as

```
.saveStatusButton {
  ...
  &:active {
    ...
  }
}
```

**Review by anna-kartynnik — APPROVED** [review]

File: frontend/src/components/Header/Header.scss (hunk 4 above)
[the remark this answers is not in the document]
Comment (ppelayo1) [reply frontend/src/components/Header/Header.scss:77]: That is pretty nifty to know. Thank you.

**Review by JMStudiosJoe — APPROVED** [review]
thank you thank you thank you @ppelayo1 & @anna-kartynnik for the great teamwork very happy to get this in!

**Pull request merged**